### ISSUE TITLE
`MappaInvokeMethodAttribute`: Add support for optional SourceProperty parameter

### ISSUE THREAD

**Issue by sanelli** [sanelli/Mappa#67]
Depends on #72.
 
- Add an optional property to `MappaInvokeMethodAttribute` to allow to define the name of the source property
- The property name can be provided as ref parameter to `TryGetStrategyForPropertyOrArgumentUsingAttributesOnMethod`
- If the name is provided it is the one that should be used by instead of the one that is matching the target name

**Comment by sanelli** [comment]
### Cursor implementation plan

Uploaded plan file: `issue_67_sourceproperty_9e12112c.plan.md`

---
---
name: Issue 67 SourceProperty
overview: Add an optional `SourcePropertyName` named parameter to `MappaInvokeMethodAttribute` so invoke-method mappings can specify a custom source property without a separate `[MappaUseProperty]`, wiring it through the generator's constructor-map strategy with full test and documentation coverage.
todos:
  - id: "1"
    content: "1. Mappa — add optional SourcePropertyName to MappaInvokeMethodAttribute with XML docs"
    status: completed
  - id: "2"
    content: "2. Mappa.Tests — add MappaInvokeMethodAttributeTests for SourcePropertyName property"
    status: completed
  - id: "3"
    content: "3. Bump MappaVersion.targets and run RebuildUponVersionChange.ps1 (Mappa consumed by Generator)"
    status: completed
  - id: "4"
    content: "4. Mappa.Generator — read SourcePropertyName in AttributeDataExtensions; resolve via ref sourceProperty in ConstructorMapStrategyDetector"
    status: completed
  - id: "5"
    content: "5. Mappa.Generator.Tests — add happy-path, warning (MP00031/MP00032), and negative tests for SourcePropertyName"
    status: completed
  - id: "6"
    content: "6. Bump MappaVersion.targets and run RebuildUponVersionChange.ps1 (Generator consumed by Samples)"
    status: completed
  - id: "7"
    content: "7. Mappa.Samples, Mappa.Samples.Aot, Mappa.Samples.Tests — add SourcePropertyName sample and runtime test"
    status: completed
  - id: "8"
    content: "8. Documentation — update mappa-attributes.md, mappa-generator-algorithm.md, Mappa/README.md, Mappa.Generator/README.md"
    status: completed
  - id: "9"
    content: "9. Run RunTestsAndReportCoverage.ps1 and fix any analyzer violations"
    status: completed
isProject: false
---

# Issue #67: MappaInvokeMethodAttribute SourcePropertyName

## Goal

[GitHub issue #67](https://github.com/sanelli/Mappa/issues/67) requests an optional source-property name on `MappaInvokeMethodAttribute`. Today, when the source property name differs from the target name, callers must combine `[MappaInvokeMethod]` with `[MappaUseProperty]` (added in #72). This feature collapses that into a single attribute via a named optional property.

**Example (target after implementation):**

```csharp
[MappaInvokeMethod(
    nameof(Target.PropertyA),
    nameof(CustomMapPropertyA),
    SourcePropertyName = nameof(Source.PropertyA))]
public partial Target Map(Source input);
```

## Current behavior

Source property resolution happens **before** attribute strategy selection in [`ConstructorMapStrategyDetector.cs`](Mappa.Generator/Algorithm/StrategyDetectors/ConstructorMapStrategyDetector.cs):

```135:168:Mappa.Generator/Algorithm/StrategyDetectors/ConstructorMapStrategyDetector.cs
string expectedSourcePropertyName;
// ... from MappaUseProperty or target name ...
PropertyMapNameMatcher.TryFindSourceProperty(..., out IPropertySymbol? sourceProperty);

if (this.TryGetStrategyForPropertyOrArgumentUsingAttributesOnMethod(
        targetParameter.Name, ..., ref sourceProperty, ...))
```

`TryGetStrategyForPropertyOrArgumentUsingAttributesOnMethod` already accepts `ref IPropertySymbol? sourceProperty`, but the `MappaInvokeMethodAttribute` branch does not modify it. `MappaAssignFromContextAttribute` is the only attribute that currently updates `ref sourceProperty`.

## Property naming

Use **`SourcePropertyName`** (not `SourceProperty`) to align with [`MappaUsePropertyAttribute`](Mappa/Attributes/MappaUsePropertyAttribute.cs).

## Architecture change

```mermaid
flowchart TD
    resolveDefault[Resolve sourceProperty by target name or MappaUseProperty]
    tryAttr[TryGetStrategyForPropertyOrArgumentUsingAttributesOnMethod]
    hasSourceName{MappaInvokeMethod.SourcePropertyName set?}
    warnMP31[Emit MP00031 if MappaUseProperty also present]
    reResolve[Re-resolve sourceProperty via PropertyMapNameMatcher exact name]
    invokeStrategy[TryGetStrategyUsingMappaInvokeMethodAttribute]
    checkMP32[Emit MP00032 if method ignores source property]

    resolveDefault --> tryAttr
    tryAttr --> hasSourceName
    hasSourceName -->|yes| warnMP31 --> reResolve --> invokeStrategy
    hasSourceName -->|no| invokeStrategy
    invokeStrategy --> checkMP32
```

## Implementation steps

### 1. Mappa — add `SourcePropertyName` optional property

**File:** [`Mappa/Attributes/MappaInvokeMethodAttribute.cs`](Mappa/Attributes/MappaInvokeMethodAttribute.cs)

- Add `public string? SourcePropertyName { get; set; }` (optional named argument; no new constructor overloads).
- Update XML docs:
  - Document usage as a named parameter on any constructor overload.
  - Update overload-selection notes: source property can come from name matching, `[MappaUseProperty]`, or `SourcePropertyName`.
  - Note that when `SourcePropertyName` is set, it overrides the default name-based match for invoke-method mapping.

### 2. Mappa.Tests — attribute surface test

**File:** new `Mappa.Tests/MappaInvokeMethodAttributeTests.cs` (mirror [`MappaStaticDependencyAttributeTests.cs`](Mappa.Tests/MappaStaticDependencyAttributeTests.cs))

- Verify `SourcePropertyName` round-trips when set via object initializer / named argument pattern.

### 3. Version bump (Mappa) and rebuild

- Increment alpha in [`MappaVersion.targets`](MappaVersion.targets) (currently `10.1.0-alpha.16` → `alpha.17`).
- Run `.\Scripts\RebuildUponVersionChange.ps1`.

### 4. Mappa.Generator — wire `SourcePropertyName` through

**4a. [`Mappa.Generator/Extensions/AttributeDataExtensions.cs`](Mappa.Generator/Extensions/AttributeDataExtensions.cs)**

Refactor `GetInvokeMethodAttributes` to iterate full `AttributeData` (not only `ConstructorArguments`), following the `GetMappaSettingsAttribute` named-argument pattern:

- Build attribute from constructor args (existing 2- and 3-arg cases).
- Read optional `SourcePropertyName` from `NamedArguments`.

**4b. [`Mappa.Generator/Algorithm/StrategyDetectors/ConstructorMapStrategyDetector.cs`](Mappa.Generator/Algorithm/StrategyDetectors/ConstructorMapStrategyDetector.cs)**

In `TryGetStrategyForPropertyOrArgumentUsingAttributesOnMethod`, `MappaInvokeMethodAttribute` case:

1. If `SourcePropertyName` is non-null/non-whitespace:
   - Call `ReportMappaUsePropertySourcePropertyWillNotBeUsedIfPresent` (MP00031) when `[MappaUseProperty]` also targets the same member — pass `nameof(MappaInvokeMethodAttribute)` as conflicting attribute (same pattern as `MappaAssignFromContext`).
   - Obtain accessible source properties from `sourceClassType` (same filter as constructor/property-init paths: non-indexer, getter accessible).
   - Re-resolve `sourceProperty` via `PropertyMapNameMatcher.TryFindSourceProperty` with `useExactNameFromAttribute: true`.
   - Pass `isConstructorParameterPath` — add a `bool` parameter to `TryGetStrategyForPropertyOrArgumentUsingAttributesOnMethod` (constructor path vs property-init path) so matching rules stay correct; update both call sites (lines ~162 and ~443).

2. Pass updated `sourceProperty` into `TryGetStrategyUsingMappaInvokeMethodAttribute`.

In `TryGetStrategyUsingMappaInvokeMethodAttribute`:

- Extend the existing MP00032 check: warn when `SourcePropertyName` is set (or `MappaUseProperty` is present) but the selected method overload does not use the source property (`InvokeMethodSourcePropertyUsage.UsesSourceProperty`).

No changes expected in [`MappaInvokeMethodAttributeStrategyBuilder.cs`](Mappa.Generator/Builders/Strategies/MappaInvokeMethodAttributeStrategyBuilder.cs) or [`MapMethodMappingAttributesValidator.cs`](Mappa.Generator/Extensions/MapMethodMappingAttributesValidator.cs) — `SourcePropertyName` is not a target name and does not need MP00033 validation.

### 5. Mappa.Generator.Tests — comprehensive coverage

Add new partial file `MappaInvokeMethodAttributeTests.SourcePropertyName.cs` (keeps existing partial class pattern).

**Happy paths (assert generated syntax + no diagnostics):**

| Scenario | Attribute overload |
|----------|-------------------|
| Empty constructor + property init | `(target, methodName, SourcePropertyName = ...)` |
| Parameterized constructor | same |
| External static method | `(target, typeof(Helper), methodName, SourcePropertyName = ...)` |
| Field/instance method | `(target, fieldName, methodName, SourcePropertyName = ...)` |
| With `MappaContext` overload that consumes source property | verify context + source property args |

Use existing tests in [`MappaInvokeMethodAttributeTests.MappaUseProperty.cs`](Mappa.Generator.Tests/MappaInvokeMethodAttributeTests.MappaUseProperty.cs) as templates — same source/target shapes, but replace `[MappaUseProperty]` with `SourcePropertyName`.

**Negative / warning paths:**

| Test | Expected |
|------|----------|
| `SourcePropertyName` + `[MappaUseProperty]` on same target | MP00031 (`MappaUsePropertySourcePropertyWillNotBeUsed`) |
| `SourcePropertyName` set, method accepts only `MappaContext` | MP00032 (`MappaUsePropertyNotUsedByInvokeMethod`) |
| `SourcePropertyName` references non-existent source property | mapping fails / no suitable strategy (follow existing no-match behavior) |
| `SourcePropertyName` set, method uses source property | no MP00032 |

Extend [`MappingAttributeConflictWarningIntegrationTests.cs`](Mappa.Generator.Tests/MappingAttributeConflictWarningIntegrationTests.cs) with one MP00031 case for `MappaInvokeMethodAttribute.SourcePropertyName` + `MappaUseProperty` if not covered in the new partial file.

### 6. Version bump (Mappa.Generator) and rebuild

- Increment alpha in `MappaVersion.targets` (`alpha.17` → `alpha.18`).
- Run `.\Scripts\RebuildUponVersionChange.ps1`.

### 7. Mappa.Samples, Mappa.Samples.Aot, Mappa.Samples.Tests

This is a **feature enhancement** (not a bug fix), so add a sample per workflow rules.

- Add a mapper method in [`Mappa.Samples/MappaInvokeMethodAttributeMappers.cs`](Mappa.Samples/MappaInvokeMethodAttributeMappers.cs) demonstrating `SourcePropertyName` replacing the two-attribute workaround.
- Register runner in [`Mappa.Samples.Aot`](Mappa.Samples.Aot) (follow existing `MappaInvokeMethodAttribute` runner pattern).
- Add runtime assertion in [`Mappa.Samples.Tests/MappaInvokeMethodAttributeMappersTests.cs`](Mappa.Samples.Tests/MappaInvokeMethodAttributeMappersTests.cs).

### 8. Documentation

- [`Documentation/mappa-attributes.md`](Documentation/mappa-attributes.md) — new subsection under `MappaInvokeMethodAttribute` documenting `SourcePropertyName`, usage examples, and interaction with `[MappaUseProperty]` (prefer `SourcePropertyName` when only invoke-method mapping needs a custom source; `MappaUseProperty` remains for non-invoke mappings).
- [`Mappa/README.md`](Mappa/README.md) — mention `SourcePropertyName` on the `MappaInvokeMethodAttribute` bullet.
- [`Documentation/mappa-generator-algorithm.md`](Documentation/mappa-generator-algorithm.md) and [`Mappa.Generator/README.md`](Mappa.Generator/README.md) — note that `[MappaInvokeMethod]` can optionally specify source property name inline.

### 9. Final validation

```powershell
.\Scripts\RunTestsAndReportCoverage.ps1
```

Ensure zero StyleCop/Sonar violations; no trailing newline after closing `}` (SA1518).

## Files touched (summary)

| Project | Key files |
|---------|-----------|
| Mappa | `MappaInvokeMethodAttribute.cs` |
| Mappa.Tests | `MappaInvokeMethodAttributeTests.cs` (new) |
| Mappa.Generator | `AttributeDataExtensions.cs`, `ConstructorMapStrategyDetector.cs` |
| Mappa.Generator.Tests | `MappaInvokeMethodAttributeTests.SourcePropertyName.cs` (new), possibly `MappingAttributeConflictWarningIntegrationTests.cs` |
| Mappa.Samples | `MappaInvokeMethodAttributeMappers.cs` |
| Mappa.Samples.Aot | runner registration |
| Mappa.Samples.Tests | `MappaInvokeMethodAttributeMappersTests.cs` |
| Documentation | `mappa-attributes.md`, `mappa-generator-algorithm.md`, `Mappa/README.md`, `Mappa.Generator/README.md` |
| Versioning | `MappaVersion.targets` (two bumps) |

## Out of scope

- No changes to `StringMapStrategyDetector` or unrelated in-progress work (#234).
- No new diagnostic IDs — reuse MP00031/MP00032.
- No branch creation, commit, or push unless instructed.

**Comment by sanelli** [comment]
### Cursor implementation plan

Uploaded plan file: `issue_67_sourceproperty_9e12112c.plan.md`

---
---
name: Issue 67 SourceProperty
overview: Add an optional `SourcePropertyName` named parameter to `MappaInvokeMethodAttribute` so invoke-method mappings can specify a custom source property without a separate `[MappaUseProperty]`, wiring it through the generator's constructor-map strategy with full test and documentation coverage.
todos:
  - id: "1"
    content: "1. Mappa — add optional SourcePropertyName to MappaInvokeMethodAttribute with XML docs"
    status: completed
  - id: "2"
    content: "2. Mappa.Tests — add MappaInvokeMethodAttributeTests for SourcePropertyName property"
    status: completed
  - id: "3"
    content: "3. Bump MappaVersion.targets and run RebuildUponVersionChange.ps1 (Mappa consumed by Generator)"
    status: completed
  - id: "4"
    content: "4. Mappa.Generator — read SourcePropertyName in AttributeDataExtensions; resolve via ref sourceProperty in ConstructorMapStrategyDetector"
    status: completed
  - id: "5"
    content: "5. Mappa.Generator.Tests — add happy-path, warning (MP00031/MP00032), and negative tests for SourcePropertyName"
    status: completed
  - id: "6"
    content: "6. Bump MappaVersion.targets and run RebuildUponVersionChange.ps1 (Generator consumed by Samples)"
    status: completed
  - id: "7"
    content: "7. Mappa.Samples, Mappa.Samples.Aot, Mappa.Samples.Tests — add SourcePropertyName sample and runtime test"
    status: completed
  - id: "8"
    content: "8. Documentation — update mappa-attributes.md, mappa-generator-algorithm.md, Mappa/README.md, Mappa.Generator/README.md"
    status: completed
  - id: "9"
    content: "9. Run RunTestsAndReportCoverage.ps1 and fix any analyzer violations"
    status: completed
isProject: false
---

# Issue #67: MappaInvokeMethodAttribute SourcePropertyName

## Goal

[GitHub issue #67](https://github.com/sanelli/Mappa/issues/67) requests an optional source-property name on `MappaInvokeMethodAttribute`. Today, when the source property name differs from the target name, callers must combine `[MappaInvokeMethod]` with `[MappaUseProperty]` (added in #72). This feature collapses that into a single attribute via a named optional property.

**Example (target after implementation):**

```csharp
[MappaInvokeMethod(
    nameof(Target.PropertyA),
    nameof(CustomMapPropertyA),
    SourcePropertyName = nameof(Source.PropertyA))]
public partial Target Map(Source input);
```

## Current behavior

Source property resolution happens **before** attribute strategy selection in [`ConstructorMapStrategyDetector.cs`](Mappa.Generator/Algorithm/StrategyDetectors/ConstructorMapStrategyDetector.cs):

```135:168:Mappa.Generator/Algorithm/StrategyDetectors/ConstructorMapStrategyDetector.cs
string expectedSourcePropertyName;
// ... from MappaUseProperty or target name ...
PropertyMapNameMatcher.TryFindSourceProperty(..., out IPropertySymbol? sourceProperty);

if (this.TryGetStrategyForPropertyOrArgumentUsingAttributesOnMethod(
        targetParameter.Name, ..., ref sourceProperty, ...))
```

`TryGetStrategyForPropertyOrArgumentUsingAttributesOnMethod` already accepts `ref IPropertySymbol? sourceProperty`, but the `MappaInvokeMethodAttribute` branch does not modify it. `MappaAssignFromContextAttribute` is the only attribute that currently updates `ref sourceProperty`.

## Property naming

Use **`SourcePropertyName`** (not `SourceProperty`) to align with [`MappaUsePropertyAttribute`](Mappa/Attributes/MappaUsePropertyAttribute.cs).

## Architecture change

```mermaid
flowchart TD
    resolveDefault[Resolve sourceProperty by target name or MappaUseProperty]
    tryAttr[TryGetStrategyForPropertyOrArgumentUsingAttributesOnMethod]
    hasSourceName{MappaInvokeMethod.SourcePropertyName set?}
    warnMP31[Emit MP00031 if MappaUseProperty also present]
    reResolve[Re-resolve sourceProperty via PropertyMapNameMatcher exact name]
    invokeStrategy[TryGetStrategyUsingMappaInvokeMethodAttribute]
    checkMP32[Emit MP00032 if method ignores source property]

    resolveDefault --> tryAttr
    tryAttr --> hasSourceName
    hasSourceName -->|yes| warnMP31 --> reResolve --> invokeStrategy
    hasSourceName -->|no| invokeStrategy
    invokeStrategy --> checkMP32
```

## Implementation steps

### 1. Mappa — add `SourcePropertyName` optional property

**File:** [`Mappa/Attributes/MappaInvokeMethodAttribute.cs`](Mappa/Attributes/MappaInvokeMethodAttribute.cs)

- Add `public string? SourcePropertyName { get; set; }` (optional named argument; no new constructor overloads).
- Update XML docs:
  - Document usage as a named parameter on any constructor overload.
  - Update overload-selection notes: source property can come from name matching, `[MappaUseProperty]`, or `SourcePropertyName`.
  - Note that when `SourcePropertyName` is set, it overrides the default name-based match for invoke-method mapping.

### 2. Mappa.Tests — attribute surface test

**File:** new `Mappa.Tests/MappaInvokeMethodAttributeTests.cs` (mirror [`MappaStaticDependencyAttributeTests.cs`](Mappa.Tests/MappaStaticDependencyAttributeTests.cs))

- Verify `SourcePropertyName` round-trips when set via object initializer / named argument pattern.

### 3. Version bump (Mappa) and rebuild

- Increment alpha in [`MappaVersion.targets`](MappaVersion.targets) (currently `10.1.0-alpha.16` → `alpha.17`).
- Run `.\Scripts\RebuildUponVersionChange.ps1`.

### 4. Mappa.Generator — wire `SourcePropertyName` through

**4a. [`Mappa.Generator/Extensions/AttributeDataExtensions.cs`](Mappa.Generator/Extensions/AttributeDataExtensions.cs)**

Refactor `GetInvokeMethodAttributes` to iterate full `AttributeData` (not only `ConstructorArguments`), following the `GetMappaSettingsAttribute` named-argument pattern:

- Build attribute from constructor args (existing 2- and 3-arg cases).
- Read optional `SourcePropertyName` from `NamedArguments`.

**4b. [`Mappa.Generator/Algorithm/StrategyDetectors/ConstructorMapStrategyDetector.cs`](Mappa.Generator/Algorithm/StrategyDetectors/ConstructorMapStrategyDetector.cs)**

In `TryGetStrategyForPropertyOrArgumentUsingAttributesOnMethod`, `MappaInvokeMethodAttribute` case:

1. If `SourcePropertyName` is non-null/non-whitespace:
   - Call `ReportMappaUsePropertySourcePropertyWillNotBeUsedIfPresent` (MP00031) when `[MappaUseProperty]` also targets the same member — pass `nameof(MappaInvokeMethodAttribute)` as conflicting attribute (same pattern as `MappaAssignFromContext`).
   - Obtain accessible source properties from `sourceClassType` (same filter as constructor/property-init paths: non-indexer, getter accessible).
   - Re-resolve `sourceProperty` via `PropertyMapNameMatcher.TryFindSourceProperty` with `useExactNameFromAttribute: true`.
   - Pass `isConstructorParameterPath` — add a `bool` parameter to `TryGetStrategyForPropertyOrArgumentUsingAttributesOnMethod` (constructor path vs property-init path) so matching rules stay correct; update both call sites (lines ~162 and ~443).

2. Pass updated `sourceProperty` into `TryGetStrategyUsingMappaInvokeMethodAttribute`.

In `TryGetStrategyUsingMappaInvokeMethodAttribute`:

- Extend the existing MP00032 check: warn when `SourcePropertyName` is set (or `MappaUseProperty` is present) but the selected method overload does not use the source property (`InvokeMethodSourcePropertyUsage.UsesSourceProperty`).

No changes expected in [`MappaInvokeMethodAttributeStrategyBuilder.cs`](Mappa.Generator/Builders/Strategies/MappaInvokeMethodAttributeStrategyBuilder.cs) or [`MapMethodMappingAttributesValidator.cs`](Mappa.Generator/Extensions/MapMethodMappingAttributesValidator.cs) — `SourcePropertyName` is not a target name and does not need MP00033 validation.

### 5. Mappa.Generator.Tests — comprehensive coverage

Add new partial file `MappaInvokeMethodAttributeTests.SourcePropertyName.cs` (keeps existing partial class pattern).

**Happy paths (assert generated syntax + no diagnostics):**

| Scenario | Attribute overload |
|----------|-------------------|
| Empty constructor + property init | `(target, methodName, SourcePropertyName = ...)` |
| Parameterized constructor | same |
| External static method | `(target, typeof(Helper), methodName, SourcePropertyName = ...)` |
| Field/instance method | `(target, fieldName, methodName, SourcePropertyName = ...)` |
| With `MappaContext` overload that consumes source property | verify context + source property args |

Use existing tests in [`MappaInvokeMethodAttributeTests.MappaUseProperty.cs`](Mappa.Generator.Tests/MappaInvokeMethodAttributeTests.MappaUseProperty.cs) as templates — same source/target shapes, but replace `[MappaUseProperty]` with `SourcePropertyName`.

**Negative / warning paths:**

| Test | Expected |
|------|----------|
| `SourcePropertyName` + `[MappaUseProperty]` on same target | MP00031 (`MappaUsePropertySourcePropertyWillNotBeUsed`) |
| `SourcePropertyName` set, method accepts only `MappaContext` | MP00032 (`MappaUsePropertyNotUsedByInvokeMethod`) |
| `SourcePropertyName` references non-existent source property | mapping fails / no suitable strategy (follow existing no-match behavior) |
| `SourcePropertyName` set, method uses source property | no MP00032 |

Extend [`MappingAttributeConflictWarningIntegrationTests.cs`](Mappa.Generator.Tests/MappingAttributeConflictWarningIntegrationTests.cs) with one MP00031 case for `MappaInvokeMethodAttribute.SourcePropertyName` + `MappaUseProperty` if not covered in the new partial file.

### 6. Version bump (Mappa.Generator) and rebuild

- Increment alpha in `MappaVersion.targets` (`alpha.17` → `alpha.18`).
- Run `.\Scripts\RebuildUponVersionChange.ps1`.

### 7. Mappa.Samples, Mappa.Samples.Aot, Mappa.Samples.Tests

This is a **feature enhancement** (not a bug fix), so add a sample per workflow rules.

- Add a mapper method in [`Mappa.Samples/MappaInvokeMethodAttributeMappers.cs`](Mappa.Samples/MappaInvokeMethodAttributeMappers.cs) demonstrating `SourcePropertyName` replacing the two-attribute workaround.
- Register runner in [`Mappa.Samples.Aot`](Mappa.Samples.Aot) (follow existing `MappaInvokeMethodAttribute` runner pattern).
- Add runtime assertion in [`Mappa.Samples.Tests/MappaInvokeMethodAttributeMappersTests.cs`](Mappa.Samples.Tests/MappaInvokeMethodAttributeMappersTests.cs).

### 8. Documentation

- [`Documentation/mappa-attributes.md`](Documentation/mappa-attributes.md) — new subsection under `MappaInvokeMethodAttribute` documenting `SourcePropertyName`, usage examples, and interaction with `[MappaUseProperty]` (prefer `SourcePropertyName` when only invoke-method mapping needs a custom source; `MappaUseProperty` remains for non-invoke mappings).
- [`Mappa/README.md`](Mappa/README.md) — mention `SourcePropertyName` on the `MappaInvokeMethodAttribute` bullet.
- [`Documentation/mappa-generator-algorithm.md`](Documentation/mappa-generator-algorithm.md) and [`Mappa.Generator/README.md`](Mappa.Generator/README.md) — note that `[MappaInvokeMethod]` can optionally specify source property name inline.

### 9. Final validation

```powershell
.\Scripts\RunTestsAndReportCoverage.ps1
```

Ensure zero StyleCop/Sonar violations; no trailing newline after closing `}` (SA1518).

## Files touched (summary)

| Project | Key files |
|---------|-----------|
| Mappa | `MappaInvokeMethodAttribute.cs` |
| Mappa.Tests | `MappaInvokeMethodAttributeTests.cs` (new) |
| Mappa.Generator | `AttributeDataExtensions.cs`, `ConstructorMapStrategyDetector.cs` |
| Mappa.Generator.Tests | `MappaInvokeMethodAttributeTests.SourcePropertyName.cs` (new), possibly `MappingAttributeConflictWarningIntegrationTests.cs` |
| Mappa.Samples | `MappaInvokeMethodAttributeMappers.cs` |
| Mappa.Samples.Aot | runner registration |
| Mappa.Samples.Tests | `MappaInvokeMethodAttributeMappersTests.cs` |
| Documentation | `mappa-attributes.md`, `mappa-generator-algorithm.md`, `Mappa/README.md`, `Mappa.Generator/README.md` |
| Versioning | `MappaVersion.targets` (two bumps) |

## Out of scope

- No changes to `StringMapStrategyDetector` or unrelated in-progress work (#234).
- No new diagnostic IDs — reuse MP00031/MP00032.
- No branch creation, commit, or push unless instructed.
